README.md
[![build status](https://secure.travis-ci.org/betamos/Node-Growler.png)](http://travis-ci.org/betamos/Node-Growler)

Node Growler
============
A [Growl][1] server for [node.js][2] which sends notifications to remote and
local Growl clients using [GNTP][3]. Could for example be used to notify you when
something happens on your node.js server, e.g. a user just logged in.

Installation
------------
	npm install growler

Dependencies
------------
* node.js >= 0.6
* [Underscore.js][4] >=1.1.5

Usage
-----
	var growler = require('growler');
	var myApp = new growler.GrowlApplication('Simple Growl App');
	myApp.setNotifications({
	  'Server Status': {}
	});
	myApp.register();
	myApp.sendNotification('Server Status', {
	  title: 'Node Growler online',
	  text: 'Wasn\'t that hard was it?'
	});

Also, check the examples directory.

Features
--------
* Custom notification icons
* Send notifications to password protected clients over the network
* Send encrypted notifications (not supported yet by Growl for OS X)

Author and license
------------------
Node Growler, Copyright 2011, Didrik Nordström

Dual licensed under the MIT or GPL Version 3 licenses.

[1]:	http://growl.info/
[2]:	http://nodejs.org/
[3]:	http://www.growlforwindows.com/gfw/help/gntp.aspx
[4]:	http://documentcloud.github.com/underscore/


CHANGELOG

2013-09-25: v0.0.2

 - Adding support for URL callbacks (on-click actions)

2011-11-25: v0.0.1

- Initial release


gntp.js
/**
 * @preserve
 *
 * NodeGrowler 0.1-alpha
 *
 * A node.js Growl server which communicates with Growl clients using GNTP 1.0
 * (Growl Notification Transport Protocol).
 *
 * Dependencies:
 * - node.js 0.6.1+
 * - Underscore.js 1.2.2+
 * - openssl determines encryptions available, only affects encrypted
 *   communication
 *
 * @author Didrik Nordström, http://betamos.se/
 *
 * @see http://nodejs.org/
 * @see http://growl.info/
 */
var util = require('util');
var net = require('net');
var crypto = require('crypto');
var SecurityGuard = require('./security-guard.js');
var _ = require('underscore');

var nl = '\r\n', nl2 = nl+nl;

/**
 * Create a Growl application. This is sort of a server which communicates
 * with a Growl client over GNTP, a TCP protocol developed specifically for
 * Growl.
 *
 * @see http://growl.info/documentation/developer/gntp.php
 *
 * @param {string} name Name of the application. Visible in the Growl client
 *   settings.
 *
 * @param {Object<string, *>=} options An object with the following keys:
 *   - {?string} hostname E.g. '123.45.67.89', 'example.com'.
 *     Defaults to 'localhost'.
 *   - {?number} port Network port number. Defaults to 23053, the GNTP port
 *   - {?number} timeout Timeout for network requests
 *   - {Buffer} icon An application icon which will appear with all
 *     notifications sent by this application. Binary image file data.
 *   - {Object.<string, (string|number|boolean|Buffer|null)>} additionalHeaders
 *     More GNTP headers to append to every query
 *
 * @param {Object<string, ?string>=} security An object with the following keys:
 *   - password Configured in the Growl client. Normally not required if on the
 *     the client is on the same machine.
 *   - hashAlgorithm Required if password given. Available algorithms:
 *     - MD5 (not very secure)
 *     - SHA1 (secure, but doesn't provide digest long enough for AES)
 *     - SHA256 (secure, default)
 *     - SHA512 (Pentagon)
 *   - encryptionsAlgorithm Note that Growl 1.3.1 for OS X does NOT support
 *     encryption yet because of stupid laws in some countries. Therefore
 *     encryption is disabled by default. In Growl for Windows it works though,
 *     and these are the available algorithms:
 *     - AES (recommended)
 *     - DES
 *     - 3DES
 */
var GrowlApplication = function(name, options, security) {
  this.name = name;
  this.options = {};

  _.defaults(this.options, options, {
    hostname: 'localhost',
    port: 23053,
    timeout: 5000, // Socket inactivity timeout
    icon: null, // Buffer,
    additionalHeaders: {},
    debugFn: function() {}
  });

  this.persistentHeaders = {
    'Origin-Software-Name': 'Node.js GNTP Library',
    'Origin-Software-Version': '1.0'
  };
  // Extend with user supplied headers
  _.defaults(this.persistentHeaders, this.options.additionalHeaders);

  // Our guard will take care of all security issues
  this.guard = new SecurityGuard(
      security.password || null,
      security.hashAlgorithm || 'SHA256',
      security.encryptionAlgorithm || null);

  this.notifications = {};
};

// Export using node.js module layer
exports.GrowlApplication = GrowlApplication;

/**
 * Add notifications to this GrowlApplication instance. This does NOT send any
 * notification, this just makes it possible to register them to the Growl
 * client (required). After that it is possible to send notifications.
 *
 * @param {Object.<string, Object>} notifications An object where keys are the
 *   names of the notifications and the values are objects with these keys:
 *   - {?string} displayName: The name of the notification, as seen in the
 *     Growl client settings. Defaults to the name of the notification.
 *   - {?boolean} enabled: If this notification should be enabled be default.
 *     Defaults to true.
 *   - {Buffer} icon: An image file buffer to display as notification icon.
 */
GrowlApplication.prototype.addNotifications = function(notifications) {
  _.each(notifications, function(options, name) {
    notifications[name] = {};
    _.defaults(notifications[name], options, {
      displayName: name, // Set display name
      enabled: true, // Enabled by default
      icon: null
    });
  });
  this.notifications = notifications;
};

/**
 * Register this application to the Growl client. All notifications that have
 * been added will be registered. If an application is registered multiple
 * times with the same name, the previous gets overwritten. It is required to
 * register before sending notifications. Registrations are persistent on the
 * client so there is no need to register if no updates to the application has
 * been made.
 *
 * @param {?function(boolean, Error=)} Always called. Possible errors are "not
 *   authorized" (usually wrong or lacking password), connection error etc.
 */
GrowlApplication.prototype.register = function(callback) {

  var headerBlocks = [{
    'Application-Name': this.name,
    'Notifications-Count': _.keys(this.notifications).length,
    'Application-Icon': this.options.icon
  }];
  _.each(this.notifications, function(options, name) {
    headerBlocks.push({
      'Notification-Name': name,
      'Notification-Display-Name': options.displayName || name,
      'Notification-Enabled': !!options.enabled,
      // There is a bug in Growl 1.3.2 which ignores this icon
      // @see GrowlApplication.prototype.sendNotification
      'Notification-Icon': options.icon
    });
  });

  this.sendQuery('REGISTER', headerBlocks, callback || function() {});
};

/**
 * Send a notification to the host.
 *
 * @param {string} name Notification name, must have already been added to the
 *   GrowlApplication object. Throws an error if it doesn't exist.
 *
 * @param {Object=} options Additional options object with the following keys:
 *   - {?string} title: Title of the message on the screen, visible to user
 *   - {?string} text: Message text, visible to the user
 *   - {?function(boolean, Error=)} callback Always called
 *   - {?boolean} sticky: Makes sure notification stays on screen until clicked
 *   - {Buffer} icon: Image file buffer
 *
 * @return {string} The randomized notification ID
 */
GrowlApplication.prototype.sendNotification = function (name, options) {

  var notification = this.notifications[name];
  if (!notification)
    throw new Error('Cannot find notification with name <'+ name +'>');

  var id = crypto.randomBytes(16).toString('hex');

  options = _.clone(options) || {};
  _.defaults(options, {
    title: notification.displayName,
    text: null,
    callback: function() {}, // Called when a response is received
    sticky: null, // Stay on screen until clicked
    priority: null, // In range [-2, 2], 2 meaning emergency
    icon: notification.icon
  });

  var headerBlocks = [{
    'Application-Name': this.name,
    'Notification-Name': name,
    'Notification-ID': id,
    'Notification-Title': options.title,
    'Notification-Text': options.text,
    'Notification-Sticky': !!options.sticky,
    'Notification-Priority': options.priority,
    'Notification-Icon': options.icon // @see GrowlApplication.prototype.register
  }];

  this.sendQuery('NOTIFY', headerBlocks, options.callback);

  return id;
};

/* PRIVATE */

/**
 * Assemble a query into a message string and attachments as buffers.
 * One possible side effect is that the headerBlocks object may be altered.
 *
 * @param {Array.<Object.<string, (string|number|boolean|Buffer|null)>>}
 *   headerBlocks An array of header blocks, where each block is an object with
 *   a string key and one of these values:
 *   - string: (GNTP <string>)
 *   - number: (GNTP <int>) Will run through parseInt to assure integer
 *   - boolean: (GNTP <boolean>) May NOT be null, see below.
 *   - Buffer: (GNTP <uniqueid>) for sending binary data, e.g. an image.
 *   - null: omits the entire header
 *
 * @return {Object} An object with the keys:
 *   - {string} message: The message, does NOT begin nor end with CRLF
 *   - {Object.<string, !Buffer>} attachments: Keys are the GNTP <uniqueid>'s,
 *     values are their corresponding buffers.
 */
GrowlApplication.prototype.assembleQuery = function(headerBlocks) {
  var self = this,
    blocks = [],
    attachments = {}; // An object with <uniqueid> as keys and buffers as values
  _.each(headerBlocks, function(header, index) {
    var lines = [];
    if (index == 0) {
      // Additional headers belongs to the first block
      _.defaults(header, self.persistentHeaders);
    }
    _.each(header, function(value, key) {
      if (typeof key != 'string' || value == null)
        return;
  
      // Special case for buffers, they will be treated as attachments
      if (Buffer.isBuffer(value)) {
        // Create an md5 hash of the buffer
        var hash = crypto.createHash('md5');
        hash.update(value);
        var digest = hash.digest('hex');
        // Add to binary
        attachments[digest] = value;
        // Point to the binary attachment and alter value
        value = 'x-growl-resource://'+ digest;
      }
  
      // Alter value so that it is completely GNTP safe
      switch (typeof value) {
      case 'string':
        break;
      case 'number':
        value = parseInt(value);
        if (isNaN(value))
          return;
        break;
      case 'boolean':
        value = value ? 'True' : 'False';
        break;
      default:
        return;
      }
      // If everything worked out, add the line
      lines.push(key +': '+ value);
    });
    blocks.push(lines.join(nl));
  });
  return {
    message: blocks.join(nl2),
    attachments: attachments
  };
};

/**
 * Retrieve the GNTP information line which often looks something like
 * "GNTP/1.0 REGISTER NONE" or similar.
 *
 * @param {string} messageType GNTP message type, e.g. 'REGISTER'
 *
 * @return {string} GNTP information line, without CRLF.
 */
GrowlApplication.prototype.getInfoLine = function(messageType) {
  var infoLine = 'GNTP/1.0 '+ messageType +' ';
  if (this.guard.encAlg) // Encryption enabled
      infoLine += this.guard.encAlg +':'+ this.guard.iv.toString('hex');
  else // No encryption
    infoLine += 'NONE';
  if (this.guard.hashAlg) // Password protection
    infoLine += ' '+ this.guard.hashAlg +':'+
                this.guard.keyHash +'.'+ this.guard.salt.toString('hex');
  // Actually, GNTP only requires the algorithm id (e.g. sha1) to be uppercase
  // but their example GNTP information lines are all uppercase so better be safe.
  return infoLine.toUpperCase();
};

/**
 * Send a query and wait for response, then call provided callback.
 * Takes care of calling appropriate security methods and encryption.
 *
 * @param {string} messageType The GNTP message type, e.g. "NOTIFY"
 *
 * @param {Array.<Object.<string, (string|number|boolean|Buffer|null)>>}
 *   headerBlocks A list of header objects.
 * @see GrowlApplication.prototype.assembleQuery
 *
 * @param {function(boolean, Error=)} callback
 *   
 */
GrowlApplication.prototype.sendQuery = function(messageType, headerBlocks, callback) {
  var self = this;
  var socket = new net.Socket();
  // Since neither Growl for OS X or Windows encrypt their responses,
  // presume plain text
  socket.setEncoding('utf8');

  // Connect
  socket.connect(this.options.port, this.options.hostname, function() {
    // Retrieve the data that shall be sent
    var data = self.assembleQuery(headerBlocks),
      infoLine = self.getInfoLine(messageType);

    self.options.debugFn('message_out', infoLine + nl + data.message);

    // Information line never encrypted
    socket.write(infoLine+nl);
    self.guard.writeSecure(socket, data.message);

    _.each(data.attachments, function(buffer, uniqueid) {
      socket.write(nl2 +'Identifier: '+ uniqueid + nl +'Length: '+ buffer.length + nl2);
      self.guard.writeSecure(socket, buffer);
    });
    socket.write(nl2);
  });

  var recieved = '';
  // Aggregate data before parsing
  socket.on('data', function(data) {
    recieved += data;
    self.options.debugFn('message_in', data.toString());
  });

  // Actively wait for remote to close socket
  socket.once('close', function(error) {
    var response = self.parseResponse(recieved);
    if (response && response.status) // All good
      callback(true);
    else if (response) { // GNTP responded with error
      var e = new Error('Host: '+
        (response.headers['Error-Description'] ? response.headers['Error-Description'] : ''));

      e.errorCode = response.headers['Error-Code'];
      callback(false, e);
    }

    else // Not even valid GNTP
      callback(false, new Error('The response was invalid GNTP.'));
  });

  // Exception management
  socket.on('error', function(exception) {
    // Could probably not connect to server
    callback(false, exception);
  });
  // We can not wait forever
  socket.setTimeout(this.options.timeout, function() {
    socket.destroy();
    callback(false, new Error('Server did not respond'));
  });
};


/**
 * Parses a raw response string into an object with information about the status
 * and other headers. Ignores lines that are not "key: value" structured.
 *
 * @param {string} data Raw response string
 *
 * @return {Object} If malformed GNTP response information line, then null
 *   or else an object with the following keys:
 *   - {boolean} status: Message type, true for '-OK', false for '-ERROR'
 *   - {Object.<string, string>} headers: An object with keys and values
 *     corresponding to the GNTP headers, e.g.
 *     - 'Response-Action': 'NOTIFY',
 *     - 'Error-Code': '402'
 */
GrowlApplication.prototype.parseResponse = function(data) {
  var lines = data.split(nl),
    matches;

  // Check for valid GNTP header
  if (!(lines.length &&
        (matches = /^GNTP\/1\.0\s\-(OK|ERROR)\sNONE\s*$/.exec(lines.shift())) &&
        matches.length == 2))
    return null; // Invalid, return null

  var status = matches[1] == 'OK';
  var headers = {};
  _(lines).chain()
    .filter(function(line) { return /^.+:\s.*$/.test(line); })
    .map(function(line) {
      // Match key: value pair
      var matches = /^(.+):\s(.*)$/.exec(line);
      if (!matches || matches.length < 3)
        throw new Error('GNTP Module internal error');

      headers[matches[1]] = matches[2];
    })
    .value(); // End chain
  return {
    status: status,
    headers: headers
  };
};


gtnp.js

var net = require('net');
var _ = require('underscore');
var crypto = require('crypto');

var nl = '\r\n', nl2 = nl+nl;

var GrowlApplication = function(applicationName, options) {
  this.name = applicationName;
  this.options = options;

  _.defaults(options, {
    hostname: 'localhost',
    port: 23053,
    timeout: 5000, // Socket inactivity timeout
    applicationIcon: null, // Buffer
    debug: false,
    additionalHeaders: ['X-Sender: Node.js GNTP Library'], // Send on every request TODO: not merged
    encryption: false,
    hashAlgorithm: 'sha1',
    password: null
  });

  this.notifications = {};
  this.debug = false;
};

exports.GrowlApplication = GrowlApplication;


/**
 * Callback will get {Bool} status, {Error} error
 */
GrowlApplication.prototype.register = function(callback) {
  var queries = [], binaryQueries = [];
  var self = this;

  var q = this.header('REGISTER');
  q.push(
    'Notifications-Count: '+ _.keys(this.notifications).length
  );

  if (Buffer.isBuffer(this.options.applicationIcon)) {
    var hash = crypto.createHash('md5');
    hash.update(this.options.applicationIcon);
    var digest = hash.digest('hex');
    q.push('Application-Icon: x-growl-resource://'+ digest);
    binaryQueries.push({
      id: digest,
      buffer: this.options.applicationIcon
    });
  }

  queries.push(self.assembleQuery(q));

  _.each(this.notifications, function(options, name) {
    var q = [
      'Notification-Name: '+ name,
      'Notification-Display-Name: '+ options.displayName || name,
      'Notification-Enabled: '+ (options.enabled ? 'True' : 'False')
    ];
    queries.push(self.assembleQuery(q));
  });
  self.sendQuery(self.assembleQueries(queries), callback || function() {}, binaryQueries);
};

GrowlApplication.prototype.sendNotification = function (name, options) {

  var not = this.notifications[name];

  _.defaults(options, {
    title: not.displayName,
    text: '',
    callback: function() {}, // Called when a response is recieved
    sticky: false, // Stay on screen until clicked
    priority: 0 // In range [-2, 2], 2 meaning emergency
  });

  var q = this.header('NOTIFY').concat(
    'Notification-Name: '+ name,
    'Notification-Title: '+ options.title,
    'Notification-Text: '+ options.text,
    'Notification-Sticky: '+ (options.sticky ? 'True' : 'False'),
    'Notification-Priority: '+ options.priority
  );
  this.sendQuery(this.assembleQuery(q), options.callback);
};

GrowlApplication.prototype.addNotifications = function(notifications) {
  _.each(notifications, function(options, name) {
    _.defaults(options, {
      displayName: name, // Set display name
      enabled: true // Enabled by default
    });
  });
  this.notifications = notifications;
  console.log(notifications);
};

/* PRIVATE STUFF */

GrowlApplication.prototype.header = function(messageType) {
  return [
    'GNTP/1.0 '+ messageType +' NONE' + this.hashHead()
  ].concat(this.options.additionalHeaders, 'Application-Name: '+ this.name);
};

GrowlApplication.prototype.hashHead = function() {
  // Need to check for type since an empty string is a valid password
  if (typeof this.options.password != 'string')
    return '';

  var salt = crypto.randomBytes(16),
    hash = crypto.createHash(this.options.hashAlgorithm),
    key, keyHash, hashHead;

  hash.update(this.options.password);
  hash.update(salt);
  // The key is pass+salt hashed
  key = hash.digest();
  // Create a new hash object
  hash = crypto.createHash(this.options.hashAlgorithm);
  // Yo dawg, we put a hash in yo hash
  hash.update(key);
  // Retrieve the final hash (digest) in hex form
  keyHash = hash.digest('hex');
  hashHead = this.options.hashAlgorithm +':'+ keyHash +'.'+ salt.toString('hex');
  // Actually, GNTP only requires the algorithm id (e.g. sha1) to be uppercase
  // but their example GNTP information lines are all uppercase so better be safe.
  return ' '+ hashHead.toUpperCase();
};

/**
 * lines: An array of lines
 */
GrowlApplication.prototype.assembleQuery = function(lines) {
  return lines.join(nl);
};

/**
 * Assemble multiple queries for sending.
 */
GrowlApplication.prototype.assembleQueries = function(queries) {
  return queries.join(nl2);
};

/**
 * Send a query and wait for response, then call callback with cb({Boolean} status, {Error|Null} err)
 */
GrowlApplication.prototype.sendQuery = function(query, cb, binaryQueries) {
  var self = this;
  var socket = new net.Socket();
  socket.setEncoding('utf8');

  if (this.options.debug)
    console.log('Sending query:\n===\n'+ query +'\n===');
  socket.connect(this.options.port, this.options.hostname, function() {
    socket.write(query);
    _.each(binaryQueries, function(bin) {
      socket.write(nl2 +'Identifier: '+ bin.id + nl +'Length: '+ bin.buffer.length + nl2);
      socket.write(bin.buffer);
    });
    socket.write(nl2);
  });
  socket.once('data', function(data) {
    socket.destroy();
    if (self.options.debug)
      console.log(data.toString());
    var response = self.parseResponse(data);
    if (response && response.status) // All good
      cb(true);
    else if (response) { // GTNP responded with error
      var e = new Error('Host: '+
        (response.headers['Error-Description'] ? response.headers['Error-Description'] : ''));

      e.errorCode = response.headers['Error-Code'];
      cb(false, e);
    }

    else // Not even valid GTNP
      cb(false, new Error('The response was invalid GTNP.'));
  });

  // Exception management
  socket.on('error', function(exception) {
    // Could probably not connect to server
    cb(false, exception);
  });
  socket.setTimeout(this.options.timeout, function() {
    socket.destroy();
    cb(false, new Error('Server did not respond'));
  });
};


/**
  Parses and returns a raw response into an object which looks like this:
  {
    status: true, // OK | ERROR => true | false
    headers: {
      'Response-Action': 'NOTIFY',
      'Error-Code': '402'
    }
  }
  or null if the GNTP information line is malformed.
  Ignores lines that are not key: value structured
 */
GrowlApplication.prototype.parseResponse = function(data) {
  var lines = data.split('\r\n'),
    matches;

  // Check for valid GNTP header
  if (!(lines.length &&
        (matches = /^GNTP\/1\.0\ \-(OK|ERROR)\ NONE$/.exec(lines.shift())) &&
        matches.length == 2))
    return null; // Invalid, return null

  var status = matches[1] == 'OK';
  var headers = {};
  _(lines).chain()
    .filter(function(line) { return /^.+:\s.*$/.test(line); })
    .map(function(line) {
      // Match key: value pair
      var matches = /^(.+):\s(.*)$/.exec(line);
      if (!matches || matches.length < 3)
        throw new Error('GTNP Module internal error')

      headers[matches[1]] = matches[2];
    })
    .value(); // End chain
  return {
    status: status,
    headers: headers
  };
};


sample.js

var gtnp = require('./gntp'),
    fs = require('fs'),
    crypto = require('crypto'),
    buffer = require('buffer'),
    net = require('net');

var myapp = new gtnp.GrowlApplication('Shit Head', {
  //hostname: '192.168.0.15',
  debug: true,
  //password: 'bacon',
  hashAlgorithm: 'sha256',
  //icon: fs.readFileSync('./volcano.jpg')
});

/*
var server = net.createServer(function(socket) {
  //socket.on('data', function() {});
});

//server.listen(1337);
*/

myapp.addNotifications({
  'Egg Shat': {},
  'Fart': {
    enabled: true,
    displayName: 'A fart just occured',
    icon: fs.readFileSync('./volcano.jpg'),
  }
});

myapp.register(function(status, err) {
  if (!status)
    throw err;

  console.log('registration accepted.');

   myapp.sendNotification('Fart', {
    callback: function(status, err) {
      if (!status)
        throw err;
      else
        console.log('Fantastic! The message was totally delivered.');
    }
  });

  myapp.sendNotification('Egg Shat', {
    text: 'Bacon egg är gott',
    sticky: true,
    icon: fs.readFileSync('ansikte.jpg'),
    callback: function(status, err) {
      if (!status)
        throw err;
      else
        console.log('Fantastic! This message was totally delivered.');
    }
  });
  
});


lib/security-guard.js
/**
 * Security management for Node Growler
 */

var crypto = require('crypto');

/*
 * Always choose an algorithm which has a digest size which is at least as long
 * as the encryption algorithm's key length. Only applies if using encryption.
 */
var hashAlgorithms = {
  'MD5': {
    name: 'md5',
    digestSize: 16
  },
  'SHA1': {
    name: 'sha1',
    digestSize: 20
  },
  'SHA256': { // Default
    name: 'sha256',
    digestSize: 32
  },
  'SHA512': {
    name: 'sha512',
    digestSize: 64
  }
};

/*
 * Note that not all of these algorithms may exist on your system
 */
var encryptionAlgorithms = {
  'AES': { // Recommended
    name: 'aes192',
    keyLength: 24,
    ivSize: 16
  },
  'DES': {
    name: 'des',
    keyLength: 8,
    ivSize: 8
  },
  '3DES': {
    name: 'des3',
    keyLength: 24,
    ivSize: 8
  }
};

var SALT_SIZE = 16;

/**
 * @constructor
 *
 * Create a security guard. Note that salt and hashes are in binary form.
 * When an instance of this class is instantiated
 * with the 'new' keyword, it will return an object with the following keys:
 * - {string|null} hashAlg Will be hashAlgorithm if password was given
 * - {Buffer=} salt Hex-encoded salt used for key (only if hashAlg)
 * - {string=} keyHash Hex-encoded key hash (only if hashAlg)
 * - {string|null} encAlg Will be encryptionAlgorithm if hashAlg and
 *   encryptionAlgorithm was given
 * - {Buffer=} iv Hex-encoded initial vector for encryption (only if encAlg)
 * - {Function} writeSecure Writes data to socket, see below.
 *
 * @param {string|null} password The password to the Growl client. If null, no
 *   security will ever work. Password CAN be empty string.
 *
 * @param {string} hashAlg Algorithm for hashing.
 *   Should be 'MD5', 'SHA1', 'SHA256' or 'SHA512'.
 *
 * @param {string|null} encAlg Encryption algorithm. Should be 'AES' or 'DES'
 *   or null if no encryption should be enabled.
 */
var SecurityGuard = function(password, hashAlgorithm, encryptionAlgorithm) {
  password = typeof password == 'string' ? password : null;

  // Need to check for type since an empty string is a valid password
  this.passwordProtection = password != null;

  // Hashing and key generation
  if (this.passwordProtection) {
    this.hashAlg = hashAlgorithm;
    //
    var saltBuffer = crypto.randomBytes(SALT_SIZE);
    this.salt = saltBuffer.toString('hex');

    var hash = crypto.createHash(hashAlgorithms[this.hashAlg].name);
  
    hash.update(password);
    hash.update(saltBuffer);
    // The key is pass+salt hashed. Don't expose it
    var key = hash.digest();
    // Create a new hash object
    hash = crypto.createHash(hashAlgorithms[this.hashAlg].name);
    // Yo dawg, we put a hash in yo hash
    hash.update(key);
    // Retrieve the final hash (digest) in hex form
    this.keyHash = hash.digest('hex');
  }
  else
    this.hashAlg = null;

  if (this.passwordProtection && encryptionAlgorithm) {
    this.encAlg = encryptionAlgorithm;
    if (encryptionAlgorithms[this.encAlg].keyLength > hashAlgorithms[this.hashAlg].digestSize)
      throw new Error('Invalid combination of hash and encryption algorithm. '+
                      this.encAlg +' needs a '+ encryptionAlgorithms[this.encAlg].keyLength +
                      'B key but '+ this.hashAlg + ' only generates a '+
                      hashAlgorithms[this.hashAlg].digestSize +'B digest.');
    // For encryption, an initial vector is required
    var ivBuffer = crypto.randomBytes(encryptionAlgorithms[this.encAlg].ivSize);
    this.iv = ivBuffer.toString('hex');
  }
  else {
    this.encAlg = null;
  }

  password = null; // Remove password from memory

  /**
   * Writes data to the socket according to the object's encryption algorithm.
   * If not encrypted it just writes the data plain.
   *
   * @param {!Socket} socket The socket to write the data to. Must be writable.
   *
   * @param {string|!Buffer} data Data to write. If string assuming UTF-8.
   */
  this.writeSecure = function(socket, data) {
    var encryption = encryptionAlgorithms[this.encAlg] || null;
    if (encryption) {
      var cipher = crypto.createCipheriv(
          encryption.name,
          key.slice(0, encryption.keyLength),
          ivBuffer.toString('binary'));
      socket.write(cipher.update(data, typeof data == 'string' ? 'utf8' : 'binary'), 'binary');
      socket.write(cipher.final(), 'binary');
    }
    else
      socket.write(data);
  };
};

// Export using node.js module layer
module.exports = SecurityGuard;


examples/advanced.js
// Shows two Growl notification on the local machine

var growler = require('growler'),
    fs = require('fs');

// Create a Growl application
var myApp = new growler.GrowlApplication('Advanced Growl App', {
  hostname: '10.211.55.4', // IP or DNS
  // port: 23053, // Default GNTP port
  // timeout: 5000, // Socket inactivity timeout
  icon: fs.readFileSync('nodejs.png'), // Buffer
  additionalHeaders: {
    'X-Foo': 'bar' // Additional GNTP headers sent on all requests
  }
}, {
  password: 'bacon', // Password is set in the Growl client settings
  // hashAlgorithm: 'SHA512', // MD5, SHA1, SHA256 (default), SHA512
  encryption: '3DES' // AES, DES or 3DES, by default no encryption
});

myApp.setNotifications({
  'Server Status': {
    // icon: null, // Overridable
    displayName: 'The Server\'s current status',
    // enabled: true // If the notification should be enabled by default
  },
  'User Notice': {} // Another notification "class"
});

// Must register to send messages
myApp.register(function(success, err) {
  if (!success)
    throw err;
  // Wait for register to complete before sending notifications
  var id = myApp.sendNotification('Server Status', {
    title: 'Node Growler online',
    text: 'Wasn\'t that hard was it?'
  });
  setTimeout(function() {
    myApp.sendNotification('Server Status', {
      title: 'Node Growler online',
      text: 'Wasn\'t that hard was it?',
      coalescingID: id
    });
  }, 500);
  myApp.sendNotification('User Notice', {
    title: 'Reactor Overheated',
    text: 'The pressure in the reactor is reaching critical levels',
    sticky: true, // Stay on screen
    priority: 2, // Critical
    icon: fs.readFileSync('radioactivity.png') // Override
  }, function(success, err) { // Callback
    if (success)
      console.log('The important notification was shown to the user');
    else {
      console.log('Warning! The important notification was not shown to the user');
      throw err;
    }
  });
});

// Print the requests and responses for debugging purposes
myApp.on('request', function(msg) {
  console.log('== SENDING ==');
  console.log(msg);
});

myApp.on('response', function(msg) {
  console.log('== RESPONSE ==');
  console.log(msg);
});
